src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs

```diff
@@ -188,5 +277,56 @@ public Task Write100ContinueAsync()
                     cancellationToken);
             }
         }
+
+        private Memory<byte> GetChunkedMemory(int sizeHint)
+        {
+            // The max size of a chunk will be the size of memory returned from the PipeWriter (today 4096)
+            // minus 5 for the max chunked prefix size and minus 2 for the chunked ending, leaving a total of
+            // 4089.
+
+            if (_currentChunkMemory.Length == 0)
+            {
+                // First time calling GetMemory
+                _currentChunkMemory = _pipeWriter.GetMemory(sizeHint);
+            }
+
+            var memoryMaxLength = _currentChunkMemory.Length - BeginChunkLengthMax - EndChunkLength;
+            if (_advancedBytesForChunk == memoryMaxLength)
+            {
+                // Chunk is completely written, commit it to the pipe so GetMemory will return a new chunk of memory.
+                CommitChunkToPipe();
+                _currentChunkMemory = _pipeWriter.GetMemory(sizeHint);
+            }
+
+            var actualMemory = _currentChunkMemory.Slice(
+                BeginChunkLengthMax + _advancedBytesForChunk,
+                memoryMaxLength - _advancedBytesForChunk);
+
+            return actualMemory;
+        }
+
+        private void CommitChunkToPipe()
+        {
+            var writer = new BufferWriter<PipeWriter>(_pipeWriter);
+
+            Debug.Assert(_advancedBytesForChunk <= _currentChunkMemory.Length);
+
+            if (_advancedBytesForChunk > 0)
+            {
+                var bytesWritten = writer.WriteBeginChunkBytes(_advancedBytesForChunk);
+                if (bytesWritten < BeginChunkLengthMax)
+                {
+                    // If the current chunk of memory isn't completely utilized, we need to copy the contents forwards.
+                    // This occurs if someone uses less than 255 bytes of the current Memory segment.
+                    // Therefore, we need to copy it forwards by either 1 or 2 bytes (depending on number of bytes)
+                    _currentChunkMemory.Slice(BeginChunkLengthMax, _advancedBytesForChunk).CopyTo(_currentChunkMemory.Slice(bytesWritten));
+                }
+
+                writer.Write(_currentChunkMemory.Slice(bytesWritten, _advancedBytesForChunk).Span);
+                writer.WriteEndChunkBytes();
+                writer.Commit();
+                _advancedBytesForChunk = 0;
```

Comment:
_currentChunkMemory needs to be sliced / advanced or dropped.

Reply:
I think it's probably best to use a bool for tracking what the current segment is rather than setting it to default accordingly.

Reply:
`_currentChunkMemoryUpdated = false;`?